examples/vision/transformer_in_transformer.py

```diff
@@ -0,0 +1,322 @@
+TNT is a novel model for modeling both patch-level and pixel-level
+representation. In each TNT block, an ***outer*** transformer block is utilized to
+process
+patch embeddings, and an ***inner***
+transformer block extracts local features from pixel embeddings. The pixel-level
+feature is projected to the space of patch embedding by a linear transformation layer
+and then added into the patch.
+This example requires TensorFlow 2.5 or higher, as well as
+[TensorFlow Addons](https://www.tensorflow.org/addons/overview) package,
+which can be installed using the following command:
+```python
+pip install -U tensorflow-addons
+```
+"""
+
+"""
+## Setup
+"""
+
+import matplotlib.pyplot as plt
+import numpy as np
+import math
+import tensorflow as tf
+import tensorflow_addons as tfa
+from tensorflow import keras
+from tensorflow.keras import layers
+from itertools import repeat
+
+
+"""
+## Prepare the data
+"""
+
+num_classes = 100
+input_shape = (32, 32, 3)
+
+(x_train, y_train), (x_test, y_test) = keras.datasets.cifar100.load_data()
+y_train = keras.utils.to_categorical(y_train, num_classes)
+y_test = keras.utils.to_categorical(y_test, num_classes)
+print(f"x_train shape: {x_train.shape} - y_train shape: {y_train.shape}")
+print(f"x_test shape: {x_test.shape} - y_test shape: {y_test.shape}")
+
+"""
+## Configure the hyperparameters
+"""
+
+weight_decay = 0.0001
+learning_rate = 0.001
+label_smoothing = 0.1
+validation_split = 0.2
+batch_size = 128
+image_size = 96  # resize images to this size
+num_epochs = 30
+
+"""
+## Use data augmentation
+"""
+
+data_augmentation = keras.Sequential(
+    [
+        layers.Normalization(),
+        layers.Resizing(image_size, image_size),
+        layers.RandomFlip("horizontal"),
+        layers.RandomRotation(factor=0.1),
+        layers.RandomContrast(factor=0.1),
+        layers.RandomZoom(height_factor=0.2, width_factor=0.2),
+    ],
+    name="data_augmentation",
+)
+# Compute the mean and the variance of the training data for normalization.
+data_augmentation.layers[0].adapt(x_train)
+
+"""
+## Implement the pixel embedding layer
+"""
+
+
+def pixel_embed(x, image_size=96, patch_size=8, in_dim=48, stride=4):
+    _, channel, height, width = x.shape
+    image_size = tuple(repeat(image_size, 2))
+    patch_size = tuple(repeat(patch_size, 2))
+    num_patches = (image_size[0] // patch_size[0]) * (image_size[1] // patch_size[1])
+    new_patch_size = [math.ceil(ps / stride) for ps in patch_size]
+    x = layers.Conv2D(in_dim, kernel_size=7, strides=stride, padding="same")(x)
+    # pixel extraction
+    x = tf.image.extract_patches(
+        images=x,
+        sizes=(1, new_patch_size[0], new_patch_size[1], 1),
+        strides=(1, new_patch_size[0], new_patch_size[1], 1),
+        rates=(1, 1, 1, 1),
+        padding="VALID",
+    )
+    x = tf.reshape(x, shape=(-1, in_dim, new_patch_size[0], new_patch_size[1]))
+    # pixel position embedding
+    pixel_position = tf.Variable(
```

Comment:
Is this supposed to be an element of trainable state? If not it should just be a static constant, not Variable. If yes, then you should be factoring this entire block as a `Layer` subclass.